Woundmed Port

gluesniffler opened the pull request:
## About the PR
This PR ports [Backmen/Ataraxia's take on Woundmed](https://github.com/Rxup/space-station-14/pull/1056). Mostly keeping it around to hold myself accountable.

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->

## Technical details
<!-- Summary of code changes for easier review. -->

## Main Todo
- [x] Fix build
- [x] Go through YML hell
- [x] Port hotfixes and a few commits from Backmen
- [x] Tweak the Health Analyzer so it can display wounds according to the design draft I had
- [x] Add the inspect interaction from ss13 when clicking on the body doll and health detailed inspect (it says what you have).
- [x] Add popups for wounds, fractures and general damage shizz.
- [x] Fix tend wounds surgery checking for wounds instead of damageablecomponent (it kinda makes sense but at the same time it doesnt?)
- [x] Allow correcting fractures via surgery.

## Stretch Goals
- [ ] Port over organ rotting
- [ ] Fix defib interactions with rotting organs
- [ ] Add some more organ damage consequences
- [ ] Add advanced surgery / surgery research
- [ ] Kill the BUI from the right click menu and instead make it click spessmen as god intended. BUI would still be available in an Operating Table console.

## What traumas are there?
The current list of traumas are:

Delimbing trauma: Self explanatory

Bone damage trauma: Self explanatory - sticks and stones break your bones, broken bone legs will be slower, and hands/arms with broken bones are prone to slipping whatever weapons you're wielding mid combat, as well as slow down your DoAfters

Organ damage: Directly damages your organ entity. For now the only organ affected by these are eyes, which means if someone shoots your head, you could go blind!

Veins damage: Unimplemented, will affect bleeding.

Nerve damage: Affects what your "perception" of pain is on your interface (are you REALLY healthy, or is it adrenaline deceiving you?)

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
- [X] I have buried myself in technical debt
- [ ] I have a design document for this
- [X] This PR is releasing next week
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
:cl: Mocho, Rxup, Trest, Rouden, CPT, gusxyz
- TEMPORARY: Bulldozed all code for TryChangeDamage involving ArmorPenetration. Needs to be reimplemented.
- remove: Changelings are no longer sever immune!
- remove: Removed mice dodging wideswing - not because i dislike it, but because the way it was implemented was supercode. ILYA WE HAVE A WHOLE FUCKING EVENT BUS, USE IT, DONT SHOVE MORE SHIT IN TRYCHANGEDAMAGE.
- tweak: Abductor organs can no longer be removed! Blame the metagamers for ruining it :trollface:
- tweak: Removed Clickpartdamage and Heavypartdamage multipliers from most weapons (esword, hyper freq blade, etc).
- tweak: Damage has been partially reworked! Now instead of your body taking damage and routing it to each part, your body's damage is a summation of the total damage taken by all your parts. This opens up avenues for medical to strategically amputate very fucked up limbs to reduce the amount of healing they need to do.
- add: The consequences of damage have also been reworked. Every instance of damage inflicts or updates a matching "wound". Which is an interface for consequences from every damage type.
- add: Added traumas! Consequences for damage which are associated with wound infliction. They are rolled for by chance based on the type of wound. For instance a slash wound would have a VERY high chance to inflict a delimbing trauma. And a blunt wound would have a very high chance to inflict organ damage and bone damage traumas. Each trauma has a minimum damage threshold that an attack must inflict in order to do its thing. (this prevents shotguns from being gigabusted :trollface:)
- add: Added a pain & consciousness system. For now its temporarily disabled, but the gist of it is that your brain is a nerveous system, which is connected to a bunch of nerves. Each body part is a nerve, and by taking damage, this contributes to the overall pain of the entity, which causes consequences at different levels. Ideally I want it to work like Monkestation's where it also doubles as a vascular system-lite with somewhat balanced values.
- tweak: Delimbing now requires two things, one being damaging a part enough to expose the bone, and the other being breaking the bone, it has 3 levels, with crunchier noises every time. You'll learn to tell them apart real fast. When a hit is done and the bone is broken, you'll roll a chance for a delimbing trauma.
- tweak: Breaking bones now doesn't outright disable your hands... however if you try to swing anything with them, you might trip and fall, or your weapons might fly off your hands. It will also make your doafters ungodly slow.
- tweak: Added groins! Still need someone to code butt attachment surgeries to them :)
- tweak: Armor is now localized! This means for instance a web vest no longer protects your entire body. Only your chest and some of your arms. RIP to the 15TC goat.
- tweak: Armor can also protect against different wounds/traumas depending on the type.
- tweak: You can no longer fully heal via topicals. Severe enough wounds will force you to get surgery.
- tweak: Your body doll might not be fully accurate and representative of the actual state of your body. Pay attention to your chat log for injuries!
- tweak: Added tourniquets. They allow you to quickly nullify some of the pain when a limb is lost, or when they are heavily bleeding.
- tweak: Barotrauma damage is now capped so that it will not automatically nuggetify you (I think).
- tweak: Explosions now dont damage your entire body! They instead pick an "area" and localize their damage on the limbs around that area.
- tweak: Surgeries now inflict pain!
- tweak: People now have a chance to scream, tweak and/or whimper when receiving damage, or in crit.
- tweak: Feet now contribute to your movement speed's values. Don't lose them!
- tweak: Delimbing now has slightly fancier animations, still kinda jank though.

gluesniffler commented:
shut the fuck up coderabbit

gluesniffler commented:
Unmarking as draft to be able to run funny tests

gluesniffler commented:
total test victory

gusxyz reviewed (dismissed):
fix the suicide tests </3

gluesniffler commented:
Alrighty, ported the remaining commits from Backmen, the current changelog so far is:

- TEMPORARY: Bulldozed all code for TryChangeDamage involving ArmorPenetration. Needs to be reimplemented.
- TEMPORARY: The paralyzed legs trait doesn't work at the moment!
- TEMPORARY: Blade heretic's sever immunity, and changeling's sever immunities do not work as of now.
- TEMPORARY: Asphyxiation and Bloodloss damage are somewhat cosmetic right now, as they mostly affect consciousness and not really inflict direct damage. Will be fixed.
- tweak: Damage has been partially reworked! Now instead of your body taking damage and routing it to each part, your body's damage is a summation of the total damage taken by all your parts. This opens up avenues for medical to strategically amputate very fucked up limbs to reduce the amount of healing they need to do.
- add: The consequences of damage have also been reworked. Every instance of damage inflicts or updates a matching "wound". Which is an interface for consequences from every damage type.
- add: Added traumas! Consequences for damage which are associated with wound infliction. They are rolled for by chance based on the type of wound. For instance a slash wound would have a VERY high chance to inflict a delimbing trauma. And a blunt wound would have a very high chance to inflict organ damage and bone damage traumas. Each trauma has a minimum damage threshold that an attack must inflict in order to do its thing. (this prevents shotguns from being gigabusted :trollface:)

(note) The current list of traumas are:

- Delimbing trauma: Self explanatory
- Bone damage trauma: Self explanatory - sticks and stones break your bones, broken bone legs will be slower, and hands/arms with broken bones are prone to slipping whatever weapons you're wielding mid combat, as well as slow down your DoAfters
- Organ damage: Directly damages your organ entity. For now the only organ affected by these are eyes, which means if someone shoots your head, you could go blind!
- Veins damage: Unimplemented, will affect bleeding.
- Nerve damage: Affects what your "perception" of pain is on your interface (are you REALLY healthy, or is it adrenaline deceiving you?)

- tweak: Feet now contribute to your movement speed's values. Don't lose them!
- tweak: Delimbing now has slightly fancier animations, still kinda jank though.
- add: Added a pain & consciousness system. For now its temporarily disabled, but the gist of it is that your brain is a nerveous system, which is connected to a bunch of nerves. Each body part is a nerve, and by taking damage, this contributes to the overall pain of the entity, which causes consequences at different levels. Ideally I want it to work like Monkestation's where it also doubles as a vascular system-lite with somewhat balanced values.
- tweak: Delimbing now requires two things, one being damaging a part enough to expose the bone, and the other being breaking the bone, it has 3 levels, with crunchier noises every time. You'll learn to tell them apart real fast. When a hit is done and the bone is broken, you'll roll a chance for a delimbing trauma.
- tweak: Breaking bones now doesn't outright disable your hands... however if you try to swing anything with them, you might trip and fall, or your weapons might fly off your hands. It will also make your doafters ungodly slow.
- tweak: Added groins! Still need someone to code butt attachment surgeries to them :)
- tweak: Armor is now localized! This means for instance a web vest no longer protects your entire body. Only your chest and some of your arms. RIP to the 15TC goat.
- tweak: Armor can also protect against different wounds/traumas depending on the type.
- tweak: You can no longer fully heal via topicals. Severe enough wounds will force you to get surgery.
- tweak:

gluesniffler commented:
Pending tasks as mentioned in the todo are mainly fixing the health analyzer to display the wounds, and making it so that the popups on any wound/trauma infliction are displayed for people. Logically also display this log while clicking on the detailed health inspect, or on the body doll status thingymabob.

I'd love to not have to release it without fixed consciousness and organ rot since those are pretty big elements for annoying medical now. But we'll see :wellregardedindividual: Apologies for the delays!

gluesniffler commented:
Update: Added the health analyzer and fixed some FTL fuckery with it. Some balance considerations I've noticed while testing are as follows:

1. Bleedout is WAY too aggressive, within like 20s of getting a critical chest wound you bleed out, should take like 10x that because replacing blood is NOT fun in this game.
2. Pain effects will be under a 10-20% chance to show the popup and have your target scream, just because :)
3. Right click damage has been indirectly nerfed as a consequence of all damage on things with a BodyComponent now being part based. trole
4. Burn damage might need a nerf? Since it burns all your parts evenly that damage adds up quick.
5. Heads need more bone health (like 2-3x) to make beheading harder.


As for code issues:

1. Going in and out of PVS range wipes the woundablevisuals
2. The animations freak the fuck out instead of properly playing like they used to.
3. Some mispredicts around the delimb noises? Stinks of unpredicted randomness.
4. Allegedly theres a ton of lag caused by shared woundables... I'd like to validate this myself before nuking all of the shared code to server/client side like how Backmen has done it.

gluesniffler commented:
> btw convert it to draft if you don't need the workflow checks, it ires me to see it in PR list :do:

i do need the checks whenever i push to it :trollface:

deltanedas commented:
dogs have human damage overlays
![10:35:38](https://github.com/user-attachments/assets/2ec5207c-da4f-4865-810a-58ea91f38b6f)

PVS error for wounds
![10:45:00](https://github.com/user-attachments/assets/feaff144-ae3e-4a3d-8a3d-439c137b3426)

gluesniffler commented:
Needs #2693 merged so that I can solve some merg conflicts and test shit :)

gusxyz commented:
Lgtm
